storage/rocksdb/clone/donor.cc

```diff
@@ -0,0 +1,1407 @@
+    assert(get_state() != donor_state::INITIAL);
+
+    myrocks::Rdb_string_writer buf;
+    metadata.serialize(buf);
+    return send_metadata(buf, cbk);
+  }
+
+#ifndef NDEBUG
+  void assert_known_file_size_same(const std::string &fn, my_off_t size) const {
+    assert(!myrocks::has_file_extension(fn, ".log"));
+    const auto name_to_id_map_itr = m_name_to_id_map.find(fn);
+    assert(name_to_id_map_itr != m_name_to_id_map.cend());
+    const auto id = name_to_id_map_itr->second;
+    const auto not_started_itr = m_not_started_files.find(id);
+    if (not_started_itr != m_not_started_files.cend()) {
+      assert(not_started_itr->second.get_size() == size);
+      return;
+    }
+    const auto in_progress_itr = m_in_progress_files.find(id);
+    if (in_progress_itr != m_in_progress_files.cend()) {
+      assert(in_progress_itr->second.get_size() == size);
+      return;
+    }
+    const auto completed_itr = m_completed_files.find(id);
+    assert(completed_itr != m_completed_files.cend());
+    assert(completed_itr->second.get_size() == size);
+  }
+#endif
+};
+
+donor::donor(const myrocks::clone::locator &l, const uchar *&loc,
+             uint &loc_len) {
+  assert(!l.has_applier_status());
+  l.serialize(m_locator_buf, loc, loc_len);
+#ifdef HAVE_PSI_INTERFACE
+  mysql_mutex_init(myrocks::clone_donor_file_metadata_mutex_key, &m_donor_mutex,
+                   MY_MUTEX_INIT_FAST);
+  mysql_cond_init(myrocks::rdb_signal_clone_reconnection_key,
+                  &m_reconnection_signal);
+#else
+  mysql_mutex_init(nullptr, &m_donor_mutex, MY_MUTEX_INIT_FAST);
+  mysql_cond_init(nullptr, &m_reconnection_signal);
+#endif
+  register_master_task();
+}
+
+donor::~donor() {
+  if (m_rdb_file_deletes_disabled) {
+    auto *const rdb = myrocks::rdb_get_rocksdb_db();
+    const auto result = rdb->EnableFileDeletions(false);
+    if (!result.ok()) {
+      myrocks::rdb_log_status_error(result,
+                                    "RocksDB file deletion re-enable failed");
+    }
+  }
+  mysql_cond_destroy(&m_reconnection_signal);
+  mysql_mutex_destroy(&m_donor_mutex);
+  session::shutdown();
+}
+
+// Returns MySQL error code
+int donor::add_checkpoint_files(bool final, std::size_t &total_new_size) {
+  mysql_mutex_assert_owner(&m_donor_mutex);
+
+  total_new_size = 0;
+
+  const auto result = myrocks::for_each_in_dir(
+      m_checkpoint.get_dir(), MY_WANT_STAT,
+      [final, this, &total_new_size](const fileinfo &f_info) {
+        std::string fn{f_info.name};
+        // The last WAL in the checkpoint WAL set will be growing until the
+        // cross-engine consistency point. To avoid tracking that, postpone
+        // copying any WALs until then.
+        if (myrocks::has_file_extension(fn, ".log")) return true;
+        if (!final && !myrocks::has_file_extension(fn, ".sst")) return true;
+        const my_off_t size = f_info.mystat->st_size;
+        add_file(std::move(fn), size);
```

Comment:
Does xtrabackup copy the files in filename order as well? I'm guessing the oldest files are the coldest ones which are least likely to change. But filename order and file/log number order are not the same since 100000.sst will be copied before 99999.sst. The order of the file probably doesn't make too much of a difference to overall copy speeds.

Reply:
As far as I can tell, XtraBackup does not do any specific ordering and so ends up copying by the filename order.

But, I am not sure that our implementation copies in the filename order, rather, it depends on how `std::unordered_map` happens to iterate. This might be considered to be an issue and we could enforce some order if we wanted.

Until rolling checkpoint, I was planning to do an optimization to send largest files first. This avoid pathological cases under multiple threads, where they all ship small files in parallel, and then one thread sends the last big file while the rest idle. I am not sure whether this still a good heuristic under rolling checkpoint. It could be if indeed as you say the oldest files are the  the ones to be least likely to change